File: storage/db-core/src/main/java/com/coffee/gu/order/OrderRepositoryImpl.java
package com.coffee.gu.order;

import com.coffee.gu.enums.EntityStatus;
import com.coffee.gu.enums.OrderState;
import com.querydsl.jpa.impl.JPAQueryFactory;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

import static com.coffee.gu.order.QOrderEntity.*;

@Repository
public class OrderRepositoryImpl implements OrderRepository {

    private final OrderJpaRepository orderJpaRepository;
    private final JPAQueryFactory jpaQueryFactory;

    public OrderRepositoryImpl(OrderJpaRepository orderJpaRepository, JPAQueryFactory jpaQueryFactory) {
        this.orderJpaRepository = orderJpaRepository;
        this.jpaQueryFactory = jpaQueryFactory;
    }

    @Override
    public Order create(Order order) {
        return orderJpaRepository.save(OrderEntity.create(order)).toModel();
    }

    @Override
    public Order save(Order order) {
        return orderJpaRepository.save(OrderEntity.from(order)).toModel();
    }

    @Override
    public List<Order> getByPrincipalKey(String principalKey) {
        return jpaQueryFactory.selectFrom(orderEntity)
                .where(
                        orderEntity.principalKey.eq(principalKey),
                        orderEntity.entityStatus.eq(EntityStatus.ACTIVE)
                )
                .fetch()
                .stream()
                .map(OrderEntity::toModel)
                .toList();
    }

    @Override
    public Optional<Order> findByOrderKey(String orderKey) {
        return Optional.ofNullable(
                jpaQueryFactory.selectFrom(orderEntity)
                        .where(
                                orderEntity.id.eq(orderKey),
                                orderEntity.entityStatus.eq(EntityStatus.ACTIVE)
                        )
                        .fetchFirst()
        ).map(OrderEntity::toModel);
    }

    @Override
    public Optional<Order> findByOrderKey(String orderKey, OrderState state) {
        return Optional.ofNullable(
                jpaQueryFactory.selectFrom(orderEntity)
                        .where(
                                orderEntity.id.eq(orderKey),
                                orderEntity.state.eq(state),
                                orderEntity.entityStatus.eq(EntityStatus.ACTIVE)
                        )
                        .fetchFirst()
        ).map(OrderEntity::toModel);
    }

}

File: storage/db-core/src/main/kotlin/com/coffee/gu/order/OrderRepositoryImpl.kt
package com.coffee.gu.order

import com.coffee.gu.CoreException
import com.coffee.gu.ErrorType
import com.coffee.gu.enums.EntityStatus
import com.coffee.gu.enums.OrderState
import com.coffee.gu.order.QOrderEntity.orderEntity
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository
import java.util.Optional

@Repository
class OrderRepositoryImpl(
    private val orderJpaRepository: OrderJpaRepository,
    private val jpaQueryFactory: JPAQueryFactory,
) : OrderRepository {

    override fun create(order: Order): Order {
        return orderJpaRepository.save(OrderEntity.create(order)).toModel()
    }

    override fun save(order: Order): Order {
        return orderJpaRepository.save(OrderEntity.from(order)).toModel()
    }

    override fun getByPrincipalKey(principalKey: String): List<Order> {
        return jpaQueryFactory
            .selectFrom(orderEntity)
            .where(
                orderEntity.principalKey.eq(principalKey),
                orderEntity.entityStatus.eq(EntityStatus.ACTIVE),
            )
            .fetch()
            .map { it.toModel() }
    }

    override fun findByOrderKey(orderKey: String): Order {
        return jpaQueryFactory
            .selectFrom(orderEntity)
            .where(
                orderEntity.orderKey.eq(orderKey),
                orderEntity.entityStatus.eq(EntityStatus.ACTIVE),
            )
            .fetchFirst()
            ?.toModel() ?: throw CoreException(ErrorType.NOT_FOUND_DATA)
    }

    override fun findByOrderKey(orderKey: String, state: OrderState): Order {
        return jpaQueryFactory
            .selectFrom(orderEntity)
            .where(
                orderEntity.orderKey.eq(orderKey),
                orderEntity.state.eq(state),
                orderEntity.entityStatus.eq(EntityStatus.ACTIVE),
            )
            .fetchFirst()
            ?.toModel() ?: throw CoreException(ErrorType.NOT_FOUND_DATA)
    }
}
